oshdb-filter/src/main/java/org/heigit/ohsome/oshdb/filter/FilterUtil.java

```diff
@@ -0,0 +1,96 @@
+package org.heigit.ohsome.oshdb.filter;
+
+import org.heigit.ohsome.oshdb.OSHDBTag;
+import org.heigit.ohsome.oshdb.osm.OSMType;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static java.util.Collections.*;
+import static java.util.Comparator.comparingLong;
+
+public class FilterUtil {
+
+    private FilterUtil() {
+        throw new IllegalStateException("utility class");
+    }
+
+    public static Map<EnumSet<OSMType>, Set<OSHDBTag>> tags(FilterExpression expression, Predicate<OSHDBTag> test) {
+        return tags(expression.normalize(), test);
+    }
+
+    public static Map<EnumSet<OSMType>, Set<OSHDBTag>> tags(List<List<Filter>> normalized, Predicate<OSHDBTag> test) {
+        return extract(normalized, filter -> {
+            if (filter instanceof TagFilterEquals tagFilter) {
+                return Stream.of(tagFilter.getTag()).filter(test);
+            } else if (filter instanceof TagFilterEqualsAny tagFilter) {
+                return Stream.of(tagFilter.getTag()).map(key -> new OSHDBTag(key.toInt(), -1)).filter(test);
+            } else if (filter instanceof TagFilterEqualsAnyOf tagFilter) {
+                return tagFilter.tags.stream().filter(test);
+            }
+            return Stream.empty();
+        });
+    }
+
+    private static <T> Map<EnumSet<OSMType>, Set<T>> extract(List<List<Filter>> normalized, Function<Filter, Stream<T>> extractor) {
+        var result = new HashMap<EnumSet<OSMType>, Set<T>>();
+        for(var group: normalized){
+            var type = EnumSet.allOf(OSMType.class);
+            var groupResult = new HashSet<T>();
+            for (var filter : group) {
+                if (filter instanceof TypeFilter typeFilter) {
+                    type = EnumSet.of(typeFilter.getType());
+                } else {
```

Comment:
I think this is missing a case: `GeometryTypeFilter`s do implicitly also filter by type (e.g. `geometry: polygon` is equivalent to something like `(type:way or (type:relation and type=multipolygon)) and <actual geometry check>`.

Something like this should work:

```suggestion
                if (filter instanceof TypeFilter typeFilter) {
                    type.retainAll(EnumSet.of(typeFilter.getType()));
                } else if (filter instanceof GeometryTypeFilter geometryFilter) {
                    type.retainAll(geometryFilter.getOSMTypes());
                } else {
```

Another question is if there are cases where we want the `extractor` to be called with `GeometryTypeFilter`s. :thinking: